### ISSUE TITLE
Severe Performance Regression in zcashd v5.0.0 on z_gettreestate RPC

### ISSUE THREAD

**Issue by adityapk00** [zcash/zcash#5957]
### Describe the issue
One of the RPC methods in zcashd has a severe performance regression over v4.6.0-2. This issue is most likely also present in v4.7.0, but I haven't tested it yet. 

### Can you reliably reproduce the issue?
LightwalletD is currently unable to run against v5.0.0 when running in production, because one of the RPC methods is taking too long to return. It is almost certainly one of these RPC methods:
- `getaddresstxids`
- `getrawtransaction`
- `getaddressbalance`
- `getaddressutxos`

What is happening is that one of these RPCs is grabbing some sort of lock inside zcashd and taking a long time to return. This causes CPU usage on one core to become 100%, and subsequent RPC calls to queue up, eventually failing with a timeout. Note that my machine has multiple cores, but zcashd consumes 100% of one core and locks up the RPCs, eventually causing timeouts. 

This is also happening only in production, which makes me think that the large number of calls coming in from Zecwallet users is causing this problem.

### The version of Zcash you were using:
v5.0.0, (but almost certainly also happens with v4.7.0)

### Machine specs:
4 core AWS t2.xlarge instance with 8GB of ram.

**Comment by adityapk00** [comment]
Update: The problematic method is `getaddresstxids`.

On v4.6.0-2, It averages ~4ms. On v5.0.0, it averages 1.6s

With v5.0.0
```
{"app":"lightwalletd","fields.time":"1.614863869s","level":"warning","method":"getaddresstxids","msg":"Timing","time":"2022-05-13T17:20:17Z"}
```

With v4.6.0-2
```
{"app":"lightwalletd","fields.time":"4.19668ms","level":"warning","method":"getaddresstxids","msg":"Timing","time":"2022-05-13T17:16:54Z"}
```

**Comment by steven-ecc** [comment]
Thanks Aditya. This is helpful.

**Comment by str4d** [comment]
The only lock I can find held within `getaddresstxids` is `cs_main` here:
https://github.com/zcash/zcash/blob/259b0e3643e8012e4bfed490e1d788c14aa4bbe3/src/rpc/misc.cpp#L833-L836

That actually seems like a bug, because `pblocktree` is accessed from `ReadAddressIndex`, and AFAICT `pblocktree` should be protected by `cs_main`.

But in any case, this is most likely caused by somewhere else holding `cs_main` for longer than anticipated, causing this and other RPC slowdowns as a side-effect.

**Comment by LarryRuane** [comment]
Aditya, can you give us the address you were testing with? I should be able to try that address here. Thanks

**Comment by steven-ecc** [comment]
@adityapk00 also could you give us the range of blocks (if any) you were using?

**Comment by ebfull** [comment]
Another question I have is whether or not the node you were running this `getaddresstxids` command on was being bombarded by lightwalletd requests at the same time.

**Comment by LarryRuane** [comment]
Aditya, I'm not able to reproduce the problem yet. Does it reproduce for you reliably? If the problem is lock contention (as str4d suspects), then the problem should not occur reliably, all the time. 

One other idea, could you try running the RPC directly using `zcash-cli` to see if it reproduces that way? Example (that will actually work on mainnet):
```
time src/zcash-cli getaddresstxids '{"addresses":["t1MkuUv2dGshjsC19SVTLCndZdQ1JhZ7N2t"]}'
```
That reports a real time for me of about 0.020 seconds (20 ms), but that includes starting up and running the `zcash-cli` executable, so that's not bad.

**Comment by nuttycom** [comment]
I've inspected every change to `cs_main` locking since v4.6.0, and I don't find any additional locking outside of the wallet operations that this could be attributed to. However, if it is a lock contention issue, it could be that some operation guarded by an existing lock is taking substantially longer than it has in the past.

**Comment by adityapk00** [comment]
It's possible that this is not a lock contention issue at all, but that the function is just taking longer to execute because of disk access or something, and that too for only some addresses. 

I'm seeing this issue in production when serving Zecwallet's users - User's wallets are querying 10s of addresses per second, and it works fine with v4.6.0-2 (CPU is ~25%), but with v5.0.0, CPU spikes to 100% within a few seconds, and then zcashd RPC calls start timing out. 

I'll try to find a case to reproduce, but I suspect this might be happening only under high load, rather than a single address.

**Comment by str4d** [comment]
Yep, I think this is disk access. I've started a new node from scratch, and at various points I can reproduce RPC lockups that coincide with `Index writing` in benchmarks:
```
2022-05-13T19:34:47.980526Z DEBUG ProcessNewBlock: bench:   - Load block from disk: 0.60ms [27.28s]
2022-05-13T19:34:47.989753Z DEBUG ProcessNewBlock: bench:       - Connect 8 transactions: 8.44ms (1.056ms/tx, 0.128ms/txin) [180.58s]
2022-05-13T19:34:47.989761Z DEBUG ProcessNewBlock: rustzcash::orchard_ffi: Orchard BatchValidator not provided, assuming disabled.
2022-05-13T19:34:47.989766Z DEBUG ProcessNewBlock: bench:     - Verify 66 txins: 8.46ms (0.128ms/txin) [180.90s]
2022-05-13T19:35:07.599115Z DEBUG ProcessNewBlock: bench:     - Index writing: 19609.33ms [272.45s]
2022-05-13T19:35:07.599145Z DEBUG ProcessNewBlock: bench:     - Callbacks: 0.04ms [0.26s]
2022-05-13T19:35:07.599164Z DEBUG ProcessNewBlock: bench:   - Connect total: 19618.63ms [472.37s]
2022-05-13T19:35:07.599485Z DEBUG ProcessNewBlock: bench:   - Flush: 0.32ms [3.99s]
2022-05-13T19:35:07.599498Z DEBUG ProcessNewBlock: bench:   - Writing chainstate: 0.02ms [0.09s]
2022-05-13T19:35:07.599615Z  INFO ProcessNewBlock: main: UpdateTip: new best hash=00000000272b682e5b4350320b8efbcd93009ba23c599ef4b7e99349de9eb1ce height=61901 bits=486575623 log2_work=48.500961 tx=304680 date=2017-02-12 18:47:30 progress=0.009248 cache=5.2MiB(409tx)
2022-05-13T19:35:07.599827Z DEBUG ProcessNewBlock: bench:   - Connect postprocess: 0.33ms [1.95s]
2022-05-13T19:35:07.599839Z DEBUG ProcessNewBlock: bench: - Connect block: 19619.91ms [505.94s]
```

**Comment by str4d** [comment]
`Index writing` corresponds to this code block:

https://github.com/zcash/zcash/blob/730d245cea160f7d78cd9a92a63809322c03fda1/src/main.cpp#L3531-L3605

**Comment by nuttycom** [comment]
@str4d since this appears to be a regression, what is the benchmark you're using to reproduce it? I can try a `git bisect` seeking where the issue was introduced if you can let me know what to run.

**Comment by str4d** [comment]
> what is the benchmark you're using to reproduce it?

`-debug=bench` (or `setlogfilter=debug` for the firehose) will start showing those in `debug.log`. Then I just started a new node with an empty datadir (so all blocks fetched from the network).

**Comment by str4d** [comment]
I'm running a mainnet `zcashd -txindex=1 -lightwalletd=1 -debug=bench` from scratch. Some examples of particularly slow block connections:
```
2022-05-13T21:39:20.753485Z DEBUG ProcessNewBlock: bench:   - Load block from disk: 0.45ms [10.60s]
2022-05-13T21:39:20.757010Z DEBUG ProcessNewBlock: bench:       - Connect 5 transactions: 2.82ms (0.564ms/tx, 0.939ms/txin) [249.50s]
2022-05-13T21:39:20.757021Z DEBUG ProcessNewBlock: bench:     - Verify 3 txins: 2.83ms (0.944ms/txin) [251.00s]
2022-05-13T21:39:49.048376Z DEBUG ProcessNewBlock: bench:     - Index writing: 28291.34ms [54.63s]
2022-05-13T21:39:49.048396Z DEBUG ProcessNewBlock: bench:       - Undo info:       0.04ms
2022-05-13T21:39:49.048401Z DEBUG ProcessNewBlock: bench:       - Tx index:        28291.26ms
2022-05-13T21:39:49.048405Z DEBUG ProcessNewBlock: bench:       - Address index:   0.04ms
2022-05-13T21:39:49.048408Z DEBUG ProcessNewBlock: bench:       - Spent index:     0.00ms
2022-05-13T21:39:49.048411Z DEBUG ProcessNewBlock: bench:       - Timestamp index: 0.00ms
2022-05-13T21:39:49.048414Z DEBUG ProcessNewBlock: bench:       - SetBestBlock:    0.00ms
2022-05-13T21:39:49.048424Z DEBUG ProcessNewBlock: bench:     - Callbacks: 0.06ms [0.24s]
2022-05-13T21:39:49.048436Z DEBUG ProcessNewBlock: bench:   - Connect total: 28294.94ms [315.51s]
2022-05-13T21:39:49.048495Z DEBUG ProcessNewBlock: bench:   - Flush: 0.06ms [16.91s]
2022-05-13T21:39:49.048501Z DEBUG ProcessNewBlock: bench:   - Writing chainstate: 0.01ms [139.03s]
2022-05-13T21:39:49.048553Z  INFO ProcessNewBlock: main: UpdateTip: new best hash=000000002844525c862c39c6f3cda243f4961e93523155c28130b0d1ad7d0be3 height=119471 bits=474058915 log2_work=49.685885 tx=630550 date=2017-05-24 03:27:56 progress=0.019138 cache=232.3MiB(18771tx)
2022-05-13T21:39:49.048734Z DEBUG ProcessNewBlock: bench:   - Connect postprocess: 0.23ms [1.35s]
2022-05-13T21:39:49.048744Z DEBUG ProcessNewBlock: bench: - Connect block: 28295.71ms [483.53s]
```

```
2022-05-13T21:40:15.868682Z DEBUG ProcessNewBlock: bench:   - Load block from disk: 1.98ms [12.01s]
2022-05-13T21:40:15.877346Z DEBUG ProcessNewBlock: bench:       - Connect 11 transactions: 7.70ms (0.700ms/tx, 0.335ms/txin) [260.96s]
2022-05-13T21:40:15.877357Z DEBUG ProcessNewBlock: bench:     - Verify 23 txins: 7.72ms (0.335ms/txin) [262.49s]
2022-05-13T21:40:47.579807Z DEBUG ProcessNewBlock: bench:     - Index writing: 31702.43ms [91.07s]
2022-05-13T21:40:47.579829Z DEBUG ProcessNewBlock: bench:       - Undo info:       0.05ms
2022-05-13T21:40:47.579833Z DEBUG ProcessNewBlock: bench:       - Tx index:        1.09ms
2022-05-13T21:40:47.579838Z DEBUG ProcessNewBlock: bench:       - Address index:   31701.29ms
2022-05-13T21:40:47.579841Z DEBUG ProcessNewBlock: bench:       - Spent index:     0.00ms
2022-05-13T21:40:47.579844Z DEBUG ProcessNewBlock: bench:       - Timestamp index: 0.00ms
2022-05-13T21:40:47.579848Z DEBUG ProcessNewBlock: bench:       - SetBestBlock:    0.00ms
2022-05-13T21:40:47.579856Z DEBUG ProcessNewBlock: bench:     - Callbacks: 0.07ms [0.29s]
2022-05-13T21:40:47.579884Z DEBUG ProcessNewBlock: bench:   - Connect total: 31711.19ms [365.12s]
2022-05-13T21:40:47.580154Z DEBUG ProcessNewBlock: bench:   - Flush: 0.27ms [17.29s]
2022-05-13T21:40:47.580166Z DEBUG ProcessNewBlock: bench:   - Writing chainstate: 0.02ms [140.27s]
2022-05-13T21:40:47.580218Z  INFO ProcessNewBlock: main: UpdateTip: new best hash=00000000065947cb11571911f36bd38d6b1d9b24b467ec24c723d5d0f1aa6400 height=122028 bits=474391846 log2_work=49.749562 tx=647843 date=2017-05-28 14:34:35 progress=0.019663 cache=212.7MiB(15031tx)
2022-05-13T21:40:47.580422Z DEBUG ProcessNewBlock: bench:   - Connect postprocess: 0.26ms [1.48s]
2022-05-13T21:40:47.580431Z DEBUG ProcessNewBlock: bench: - Connect block: 31713.72ms [536.32s]
```

**Comment by str4d** [comment]
This is also partly a `lightwalletd` problem. It's not great that `zcashd`'s global locking (inherited from upstream) can cause these kinds of slowdowns. However, `lightwalletd` was originally designed as a caching layer so that it could scale to large numbers of clients, and not be overly affected by this kind of issue. The fact that for the newer transparent gRPC methods it only acts as a direct passthrough to the `zcashd` RPC, without any caching or timeout handling, is a problem that should be solved there. In particular, if `lightwalletd` built and maintained the address index itself (from the same stream of blocks that it builds its compact block cache), then it would be completely impervious to this kind of problem.

**Comment by LarryRuane** [comment]
Great idea, str4d; I just created a ticket: https://github.com/zcash/lightwalletd/issues/387

**Comment by str4d** [comment]
Here's the patch I created to get more details on per-index writing costs (shown with `-debug=bench`):
```diff
diff --git a/src/main.cpp b/src/main.cpp
index 18cd8ca65e..2fc4d221f0 100644
--- a/src/main.cpp
+++ b/src/main.cpp
@@ -3556,10 +3556,12 @@ bool ConnectBlock(const CBlock& block, CValidationState& state, CBlockIndex* pin
         pindex->RaiseValidity(BLOCK_VALID_SCRIPTS);
         setDirtyBlockIndex.insert(pindex);
     }
+    int64_t nTime3a = GetTimeMicros();
 
     if (fTxIndex)
         if (!pblocktree->WriteTxIndex(vPos))
             return AbortNode(state, "Failed to write transaction index");
+    int64_t nTime3b = GetTimeMicros();
 
     // START insightexplorer
     if (fAddressIndex) {
@@ -3570,11 +3572,13 @@ bool ConnectBlock(const CBlock& block, CValidationState& state, CBlockIndex* pin
             return AbortNode(state, "Failed to write address unspent index");
         }
     }
+    int64_t nTime3c = GetTimeMicros();
     if (fSpentIndex) {
         if (!pblocktree->UpdateSpentIndex(spentIndex)) {
             return AbortNode(state, "Failed to write spent index");
         }
     }
+    int64_t nTime3d = GetTimeMicros();
     if (fTimestampIndex) {
         unsigned int logicalTS = pindex->nTime;
         unsigned int prevLogicalTS = 0;
@@ -3595,6 +3599,7 @@ bool ConnectBlock(const CBlock& block, CValidationState& state, CBlockIndex* pin
         if (!pblocktree->WriteTimestampBlockIndex(CTimestampBlockIndexKey(pindex->GetBlockHash()), CTimestampBlockIndexValue(logicalTS)))
             return AbortNode(state, "Failed to write blockhash index");
     }
+    int64_t nTime3e = GetTimeMicros();
     // END insightexplorer
 
     // add this block to the view's block chain
@@ -3602,6 +3607,12 @@ bool ConnectBlock(const CBlock& block, CValidationState& state, CBlockIndex* pin
 
     int64_t nTime3 = GetTimeMicros(); nTimeIndex += nTime3 - nTime2;
     LogPrint("bench", "    - Index writing: %.2fms [%.2fs]\n", 0.001 * (nTime3 - nTime2), nTimeIndex * 0.000001);
+    LogPrint("bench", "      - Undo info:       %.2fms\n", 0.001 * (nTime3a - nTime2));
+    LogPrint("bench", "      - Tx index:        %.2fms\n", 0.001 * (nTime3b - nTime3a));
+    LogPrint("bench", "      - Address index:   %.2fms\n", 0.001 * (nTime3c - nTime3b));
+    LogPrint("bench", "      - Spent index:     %.2fms\n", 0.001 * (nTime3d - nTime3c));
+    LogPrint("bench", "      - Timestamp index: %.2fms\n", 0.001 * (nTime3e - nTime3d));
+    LogPrint("bench", "      - SetBestBlock:    %.2fms\n", 0.001 * (nTime3 - nTime3e));
 
     // Watch for changes to the previous coinbase transaction.
     static uint256 hashPrevBestCoinBase;
```

**Comment by str4d** [comment]
Assuming that disk IO from writing indices is indeed the cause of @adityapk00's RPC issues, there's very little we can do to fix it, because there is very little between these timings and LevelDB:

https://github.com/zcash/zcash/blob/730d245cea160f7d78cd9a92a63809322c03fda1/src/txdb.cpp#L409-L414

https://github.com/zcash/zcash/blob/730d245cea160f7d78cd9a92a63809322c03fda1/src/txdb.cpp#L451-L456

https://github.com/zcash/zcash/blob/730d245cea160f7d78cd9a92a63809322c03fda1/src/txdb.cpp#L418-L429

https://github.com/zcash/zcash/blob/730d245cea160f7d78cd9a92a63809322c03fda1/src/dbwrapper.cpp#L73-L78

Perhaps upgrading LevelDB will help (#5407), but that requires four other backports to land first.

**Comment by adityapk00** [comment]
Is the Index writing new code? Was it also present in v4.6.0-2? If it's new, can we disable it?

**Comment by str4d** [comment]
It is not new:
- `WriteTxIndex` has been in the codebase for at least 9 years, and corresponds to `-txindex=1` flag.
- `WriteAddressIndex` and `UpdateAddressUnspentIndex` were added 3 years ago in #3837 (to merge the Insight block explorer patchset into `zcashd`), and exposed via `-lightwalletd=1` in #4402 two years ago.

**Comment by adityapk00** [comment]
Hmm, I'm still struggling to figure out what exactly is happening to my `zcashd`. 

On my prod server, I can switch back and forth with `v4.6.0-2` and `v5.0.0`, and this issue reproduces 100% of the time. 

With `v4.6.0-2`, everything works great. As soon as I shut it down and start `v5.0.0`, LightwalletD starts returning errors 15 seconds in, clients get timeout errors and CPU gets stuck at 100%. Switch back to `v4.6.0-2`, everything is back to normal.

**Comment by adityapk00** [comment]
Does `v5.0.0` build new indexes when it upgrades from a datadir for `v4.6.0-2`? Does it spend some time after startup updating new Index? Maybe I just need to wait for the Indexes to finish creating?

**Comment by str4d** [comment]
There were no changes to how indices are loaded at start; the only related change I can think of is that `-reindex` now implies `-rescan`, which should be irrelevant here. In any case, if there _were_ index-related changes, they should block the RPC (like happens during reindexing).

@adityapk00 can you run my above patch alongside `-debug` and see if you can spot points in the `debug.log` that match up with RPC errors?

**Comment by str4d** [comment]
Other ideas for what might be contributing here between v4.6.0-2 and v5.0.0 (I'll update this as I think of more):
- Dependency updates
  - We migrated from Boost 1.78 to 1.79.
  - We updated Rust dependencies; that should not affect index writing, but could potentially affect how long consensus checks take to run in some pathological case?
  - We now use the `secp256k1` Rust crate for parsing addresses, but that should be using the same `libsecp256k1` symbols we've already been using for consensus.

**Comment by str4d** [comment]
@adityapk00 once you're done looking for `debug.log` entries that match RPC lockups, I'd recommend rebuilding `4.6.0-2` yourself with a delayed EoS halt, so that `lightwalletd` is running without issues in the interim. Then we can continue looking for 5.0.0 solutions before NU5 activates.

**Comment by adityapk00** [comment]
@str4d - Ok. I rebuilt `4.6.0-2` with 2 additional weeks for deprecation, so I think I bought some time.

**Comment by adityapk00** [comment]
Update: I built a caching layer in LightwalletD that caches the t-address -> txid mapping as it is processing blocks, and serves `getaddresstxids` from this cache. This seems to have made `zcashd` happy, and now `LightwalletD` is running against `5.0.0`.

**Comment by steven-ecc** [comment]
Nice.  Great job!

**Comment by LarryRuane** [comment]
Aditya, would you like us to merge your changes into our lightwalletd repo? If so, can you provide a patch or a branch or a PR?

**Comment by adityapk00** [comment]
Yes, the implementation is a bit of a mess right now, since I made it in a hurry, let me clean it up a bit.

**Comment by LarryRuane** [comment]
> that caches the t-address -> txid mapping as it is processing blocks

What does your side do with the list of txids returned from the `GetTaddressTxids` gRPC? The only thing I can think of that it can do -- correct me if I'm wrong! -- is call the `GetTransaction(txid)` gRPC on each of them. This gRPC synchronously calls zcashd's `getrawtransaction` to get the full transaction (that is, lightwalletd does not cache full transactions). 

So wouldn't _only_ caching the mapping from taddr to list of txids (so that `GetTaddressTxids` doesn't need to call zcashd's `getaddresstxids`) not be much of an improvement? Suppose `GetTaddressTxids` returns 20 txids for a particular address (this seems like it may be typical). Then caching the taddr -> txid mapping would reduce the number of zcashd rpc calls from 21 to 20, not much of an improvement.

So do we need to also cache all transactions (in full format, not compact format) that involve at least one taddr? If so, it can be done, but I'm concerned because _nearly all_ transactions have at least one taddr -- isn't the fraction of fully-shielded transactions only a few percent of all transactions? 

So we would end up caching nearly the entire blockchain, which is about 30gb (serialized, this is the sum of the `.zcash/block/blk?????.dat` files for mainnet). Compare that size to the current lightwalletd storage requirement, 300mb. That would be a 100x increase in storage requirements. 

Another problem is that the time lightwalletd would need to do the initial sync would be similar to the time for zcashd to do an IBD -- it may even be greater, because zcashd downloads blocks in parallel (which lightwalletd could do also, but that adds significant complexity, and it would be talking to a single zcashd so probably wouldn't help much).

It would really help to know what performance requirements you (and other wallets) have for the `GetTransaction` gRPC. Is it acceptable to call zcashd's `getrawtransaction` on each request?

**Comment by adityapk00** [comment]
Larry, you are largely correct, and I don't think this caching is going to be a good solution. 

The way I had implemented it was to cache the address -> txid mapping while processing blocks - So it does increase startup time significantly (order of hours) and takes an additional ~25G of disk space. Not great. 

Then, the `GetTaddressTxids` is served from the cache, preventing any calls to `zcashd`'s `getaddresstxids`. It does call `getrawtransaction` for each of the served Txids, but only when a client requests it, so it's not so bad. 
While this solved the immediate perf problems, it creates a lot of new ones (Additionally, my implementation has all kinds of bugs and edge cases - Reorgs, block ranges etc...). 

Another issue is that I think the performance regression also exists on other RPCs - `getaddressutxos` for example, and caching that will be very expensive indeed. 

I don't have a good solution for these issues yet...

**Comment by LarryRuane** [comment]
That makes sense, thanks. I want to correct one thing I said [above](https://github.com/zcash/zcash/issues/5957#issuecomment-1127823733), the initial sync time wouldn't be nearly as bad as the zcashd IBD, because lightwalletd doesn't have to validate each transaction and each block. It's already the case that the initial lightwalletd sync moves about 30gb over the (local) network (in the form of zcashd `getblock` calls to fetch each block), but much less needs to be written to disk (as I said, only about 300mb). Adding any sort of caching (address -> txid mapping, or full transactions) will increase the amount of disk write data (by a lot), and how much of a performance impact that has depends on things like whether you're using SSD or spinning disk. The last time I checked (a few months ago), lightwalletd initial sync (mainnet) took about an hour on my laptop (which has SSD).

**Comment by adityapk00** [comment]
After, much investigation, I think my new hypothesis is:

The problematic function is `z_gettreestate`. It is substantially slower in v5.0.0 than in v4.6.0-2

```
time curl --user foo:bar --data-binary '{"jsonrpc": "1.0", "id":"curltest", "method": "z_gettreestate", "params": ["1669880"] }' -H 'content-type: text/plain;' http://127.0.0.1:8232/
```

Takes ~3s on v5.0.0 vs <100ms on v4.6.0-2

**Comment by adityapk00** [comment]
I'm now 99% sure `z_gettreestate` is the problem. Deleting these lines (https://github.com/zcash/zcash/blob/09b02bedc9f90d16e64311bcc5ddb013dfbd33a4/src/rpc/blockchain.cpp#L1340-L1362) from `zcashd` `v5.0.0` source and recompiling fixes all issues with zcashd performance. 

I think getting the orchard treestate is very slow for some reason, which is causing cascading performance problems.

**Comment by nuttycom** [comment]
Yup, this is a bug - it's traversing the whole block index in this call, because it's before NU5 activation.

**Comment by nuttycom** [comment]
@adityapk00 thanks for digging into that; please check if #5963 resolves the problems for you.

**Comment by adityapk00** [comment]
Added one comment, but looks good to me!

**Comment by LarryRuane** [comment]
Aditya, I know you have some code for this, but wanted to explore how the needed caching could be accomplished by extending the `blocks` file (that currently contains only compact blocks). I especially wanted to figure out an efficient in-memory representation, and here's a small test / demo / POC program I came up with:

<details>
<summary>Click to expand test program</summary>

```
#!/bin/env python3
#
# We'd like to speed up the GetTaddressTxids gRPC by not
# requiring it to submit a getaddresstxids RPC to zcashd each
# time. This program investigates which data structures we
# might use in lightwalletd to accomplish this. One way is
# use sqlite3, but we want to minimize both disk and memory
# (this is *light*walletd after all). So here's an idea
# that extends the existing compact-block caching (which just
# uses two flat files) to accomplish this. This is not
# necessarily the way it will be implemented, it's just an
# exploration.
#
# The GetTaddressTxids gRPC specifies a t-address, and we must
# return the txids of all transactions that contain that address
# at least once as either an input or an output.
#
# The obvious way to implement this is to add data after each
# compact block in the `blocks` flat file. (This will require
# the blocks file to be recreated once on upgrade, but that's
# not too bad.) The additional data will consist of a list of
# (txid, list of addresses) pairs. This can easily be generated
# while downloading each block, and written to disk immediately
# after writing the compact representation of the block.
#
# The txids that appear here correspond to transactions that
# have at least one transparent address, that is, are not fully
# shielded. Each txid is followed by a list of t-addrs that
# are contained in that transaction as either an input or output
# (with duplicates removed, if any).
# 
# Here's a representation of the blocks file (only the compact
# blocks exist in this file today):
#
#   compact-block-1
#   txid1, taddr1, taddr2, ...
#   txid2, taddr3, taddr2, ...
#   txid3, taddr1, taddr4, ...
#   ...
#   compact-block-2
#   txid4, taddr5, taddr2, ...
#   txid5, taddr4, taddr9, ...
#   ...
#
# Both the txid and t-addresses are strings. Note that the txids
# are unique (duplicate transactions are never allowed across
# the entire blockchain), but the addresses are not unique.
#
# For in-memory representation, the most straightforward way
# would be to have a map from address to list-of-txids. But
# this would consume a lot of memory, since txids are 32 bytes
# and it's common for an address to appear in many transactions.
#
# So the approach we'll explore here is to have a map from
# address to list-of-indices, and an array (list in Python,
# slice in Go) of txids. To generate the list of txids for a
# given address, we look up the address in the map, and then
# use each index in data part of that map entry to look up
# the txid, and then return this list of txids. This will
# require only 4 bytes for each transaction that an address
# is part of, the address itself (the map key, 35 bytes), and
# the array of (unique) txids (32 bytes each).

## Global variables (not persisted to disk):

# list of unique txids that have one or more transparent addresses
# (these will be in order of blocks and transactions within a block)
g_txid = []

# map t-address to list of indices into g_txid[]
g_addr_to_txid = {}

def test_main():
    # during startup, add an entry that was just read from disk
    # (a list of these follows each compact block)
    add('txid1', ['addr1', 'addr2', 'addr3'])
    add('txid2', ['addr1', 'addr4'])
    GetTaddressTxids('addr1')
    GetTaddressTxids('no-such-address')
    GetTaddressTxids('addr4')

# This models what happens during startup, while reading the
# new entries that will follow the compact blocks on disk.
def add(txid, addr_list):
    print(f"debug: add({txid=} {addr_list=})")
    next_index = len(g_txid)
    g_txid.append(txid)
    for a in addr_list:
        if a not in g_addr_to_txid: g_addr_to_txid[a] = []
        g_addr_to_txid[a].append(next_index)

# This models the GetTaddressTxids gRPC.
# Given an address, return the list of corresponding txids
def GetTaddressTxids(addr):
    print(f"debug: GetTaddressTxids({addr=})")
    txids = []
    if addr in g_addr_to_txid:
        txids = [g_txid[txid] for txid in g_addr_to_txid[addr]]
    print(f"  result: {txids=}")

if __name__ == '__main__': test_main()
```

</details>

Here is the result of running the above:

<details>
<summary>Click to expand results</summary>

```
debug: add(txid='txid1' addr_list=['addr1', 'addr2', 'addr3'])
debug: add(txid='txid2' addr_list=['addr1', 'addr4'])
debug: GetTaddressTxids(addr='addr1')
  result: txids=['txid1', 'txid2']
debug: GetTaddressTxids(addr='no-such-address')
  result: txids=[]
debug: GetTaddressTxids(addr='addr4')
  result: txids=['txid2']
```
</details>

**Comment by LarryRuane** [comment]
Disregard my most recent [comment](https://github.com/zcash/zcash/issues/5957#issuecomment-1128137622), I hadn't seen the rest of the discussion. It seems that the performance of `GetTaddressTxids` wasn't the problem; if and when it is, we can reconsider improvements to it.

**Comment by adityapk00** [comment]
@nuttycom Quick question 

I saw you changed the bug fix for this a bit, and will it work correctly even if NU5 is active but the treestate is requested for a historical (pre NU5) block? This scenario (requesting treestate for historical blocks) is done by Zecwallet during blazesync for verification.

**Comment by lia174** [comment]
Hello. my iPhone updated zecwallet lite to version v5.0  since then i can't open my zecwallet lite anymore, i need to do a coin transfer but the app won't open, what should i do, how can i get access to my zec?

**Comment by steven-ecc** [comment]
Hello. This is the repository for zcashd, the full node implementation. You might try opening a ticket [here](https://github.com/ZcashFoundation/zecwallet-lite) which is the repo for zecwallet-lite.
